### ISSUE TITLE
Fluent UI v9: NextJS 13 + App router support

### ISSUE THREAD

**Issue by sopranopillow** [microsoft/fluentui#29361]
## What's the problem?

Currently we are facing issues with NextJS 13 due to its new App router using Server Components.

### What are [React Server components](https://nextjs.org/docs/app/building-your-application/rendering/server-components)?

React Server Components(RSC) are a new type of components that React has made available as unstable that allow caching of components and have many benefits. Note this is not the same as Server Side Rendering (SSR), but a similar approach of relaying work to the server.

When using RSC, React is able to run specific part of your code in the server that would normally take many resources on the client. This includes Data Fetching, server only bundles(if you only need the fetching API on the server, why add it to the client?), seach engine optimization, and the new Streaming workflow where you split chunks of UI to be rendered on the server or on the client. 

While this sounds like a huge win, this isn't stable as of yet, which brings us to the problem mentioned above. NextJS 13 has released as stable version 13 which takes advantage of RSCs. The important fact is not that RSCs are unstable, but the experience of using our library in NextJS 13, a very popular framework in the React community.

#### React Context

RSCs do not play well with React Context, as we know context cannot be used in server side. This is unfortunate since we rely heavily on our context solution for our library.

### Griffel

Part of the problem with Griffel comes with Context, since it's used in the Renderer provider. Another issue is that our makeStylesCalls are supposed to be run in the client side and so we run into an issue of them being run in server side. It is important to note here that CSS-in-JS libraries such as emotion do not support streaming and this is not up to the library to fix as it hasn't even been published as stable.

### Fluent UI v9

Since v9 uses Griffel and it uses Context extensively in its providers and component, we are also bound to run this code in the client side. While this is a fix, we are restricting client vs server code to only client and the user will have to extract the logic that can be run in server and pass it down to the components.

|   Server    | ---------- | Client  |
|:-----------:| ---------- |:-------:|
|   getData   | <--------- | Persona |
| provideData | ---------> | Persona |

### Solition 1: Re-export everything with "use client"

Add an export map to create a NextJS appdir specific export which has the `"use client"` directive. This will mean all our v9 library will run as client components.

This solution still requires setting up a registry with `useServerInsertedHTML` for Griffel to work correctly.

👍 User gets client directive out of the box.

👎 We need to add exports in both Fluent UI v9 and Griffel. 

### Solution 2: Update docs on how to make it work

Update the docs and show how to make NextJS 13 work with our library. This is basically the same with solution 1 except that the user will have to actively move our components to a file that has the `"use client"` directive rather than having it out of the box.

Here's an example of how this could be achieved: https://stackblitz.com/edit/stackblitz-starters-wmbhn4?file=app%2FFluentV9Registry.tsx

👍 No need for changes in our codebases.

👎 It's a little more work for the user to setup the files in a way they can add the client directive.

### Remarks

We will have to run our library on the client side regardless of the solution. Until we have proper guidance from React, we should not commit to an opinionated solution or "hacks" to make it work. Both these solutions unblock NextJS users and do not force opinions.

This issue is made to track the decision of this feature request: #28196

**Comment by AkiraVoid** [comment]
I must say there may be some SSR issues with solution 2:

1. CSS variables will not be defined before the first client hydration.
2. React keeps saying client HTML tags not match with server's by some reasons.
3. With the limitations of `useServerInsertedHTML`, the style tags of Fluent UI will be inserted both in head and body, which may influence SEO.

I don't know if these issues happen in just some certain versions or not.

**Comment by elliot-huffman** [comment]
@AkiraVoid is right, `solution 2` causes SSR hydration issues.

I think that @layershifter's recommendation is the best, add 'use client' to the top of each file to force client-side rendering. This will solve a variety of hydration issues that occur when using the library the way it is currently set up and documented (see these [docs for details](https://react.fluentui.dev/?path=/docs/concepts-developer-server-side-rendering-next-js-setup--page)) on what is causing hydration issues.

Here is an example of hydration issues as @AkiraVoid references:
![Screenshot of Next.JS hydration issues](https://github.com/microsoft/fluentui/assets/5030490/d063a944-1bb5-46ae-bb69-adc16808fb7c)

**Comment by AkiraVoid** [comment]
@elliot-huffman Have you upgraded   both next.js and Fluent UI to latest version? I encountered this issue before and it seems to be caused by `FluentProvider`, but in latest version it disappeared and another problem (Fluent complaint duplicated ID, which may because client hydration reset the ID counter) occurred. 

May you comment all Fluent-related components one by one, to confirm if it really is caused by `FluentProvider`?

**Comment by elliot-huffman** [comment]
> @elliot-huffman Have you upgraded both next.js and Fluent UI to latest version? I encountered this issue before and it seems to be caused by `FluentProvider`, but in latest version it disappeared and another problem (Fluent complaint duplicated ID, which may because client hydration reset the ID counter) occurred.
> 
> May you comment all Fluent-related components one by one, to confirm if it really is caused by `FluentProvider`?

I am using the latest as of today. (I run NCU daily). I notice if I refresh the page, the error disappears but if I refresh again, the error is present.

My project is pure client-side rendering, this error is only present when running the project in dev mode (which runs it in SSR).
When I run the build command, no issues are present since it is a CSR only build.

**Comment by AkiraVoid** [comment]
@sopranopillow Any updates?

**Comment by leduyquang753** [comment]
Even after putting all components into `"use client";` files there are still hydration errors as soon as I use things like `TabList`.

```
Error: Hydration failed because the initial UI does not match what was rendered on the server.
Warning: Expected server HTML to contain a matching <button> in <div>.
```

**Comment by layershifter** [comment]
> there are still hydration errors as soon as I use things like `TabList`.
> 
> ```
> Error: Hydration failed because the initial UI does not match what was rendered on the server.
> Warning: Expected server HTML to contain a matching <button> in <div>.
> ```

This looks like a bug, please report a separate issue.

**Comment by elliot-huffman** [comment]
> > there are still hydration errors as soon as I use things like `TabList`.
> > 
> > ```
> > Error: Hydration failed because the initial UI does not match what was rendered on the server.
> > Warning: Expected server HTML to contain a matching <button> in <div>.
> > ```
> 
> This looks like a bug, please report a separate issue.

I am not certain that this is a separate issue, how would you go about proving this behavior as a distinctly different issue?

From my point of view, this error is because the original post's feature has not been met...

**Comment by layershifter** [comment]
> I am not certain that this is a separate issue, how would you go about proving this behavior as a distinctly different issue?

It's a different issue as `Button`s can be rendered without warnings:

<img width="852" alt="image" src="https://github.com/microsoft/fluentui/assets/14183168/f78bc3c1-2bb1-4577-a35b-5c98e3d37d19">

(I don't count warnings about the provider as it should be fixed separately)

Once `Tab`s are used:

<img width="523" alt="image" src="https://github.com/microsoft/fluentui/assets/14183168/d606a69c-0c93-4c24-9642-c4cde34e107d">

(which seems to be a Tabster issue)

I was not able to reproduce `Warning: Expected server HTML to contain a matching <button> in <div>.`.

Sandbox: https://stackblitz.com/edit/stackblitz-starters-o9rw51

**Comment by elliot-huffman** [comment]
Good point, I agree. This does need to be brought out into a separate issue, my bad.

**Comment by KocakAli** [comment]
Hey, I am using next js 14 and fluent ui 8.113 and I get the error warning: prop `classname` did not match server: "css-149" client: "css-151"

Is there a solution to this problem because I can't find a solution I'm using typescript and app router.

**Comment by AkiraVoid** [comment]
@KocakAli Please try to use FluentUI 9 or open another issue since this issue is for FluentUI 9.

**Comment by elliot-huffman** [comment]
You should be using v9.
V8 is basically deprecated and it's highly recommended to be using v9

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by sopranopillow** [comment]
> @sopranopillow Esteban Munoz Facusse FTE Any updates?

Hey @AkiraVoid, sorry for the long wait, I'm currently working on the swc plugin. The plugin is partially done, I'm in the process of connecting it to NextJS to provide the docs for how to use it. Here's the repo that will house the plugin: https://github.com/sopranopillow/fluentui-nextjs-appdir-plugin

**Comment by sopranopillow** [comment]
Closing this issue to keep conversation in #28196

**Comment by amroncz** [comment]
> > @sopranopillow Esteban Munoz Facusse FTE Any updates?
> 
> Hey @AkiraVoid, sorry for the long wait, I'm currently working on the swc plugin. The plugin is partially done, I'm in the process of connecting it to NextJS to provide the docs for how to use it. Here's the repo that will house the plugin: https://github.com/sopranopillow/fluentui-nextjs-appdir-plugin

Is the plugin already finished and working?
Unfortunately, it does not work for me in a completely new project with the latest version of next.js and the latest version of fluentui.

Do you have an example project?

Does the plugin work with next.config.mjs or do I have to rename the file to .js?

Btw: The swc plugin swc-useclient (https://github.com/coder-xiaotian/swc-useclient) doesn't get fluentui to work for me either.

Übersetzt mit DeepL https://www.deepl.com/app/?utm_source=android&utm_medium=app&utm_campaign=share-translation

**Comment by sopranopillow** [comment]
> Is the plugin already finished and working? Unfortunately, it does not work for me in a completely new project with the latest version of next.js and the latest version of fluentui.
> 
> Do you have an example project?
> 
> Does the plugin work with next.config.mjs or do I have to rename the file to .js?

@amroncz Will investigate and will try to create a base stackblitz with a working setup. Currently don't have the bandwidth to look into it, but once I get a change I'll update the issue #30766. Would be nice if you could add your log to the issue I linked, thanks!

**Comment by amroncz** [comment]
> > Is the plugin already finished and working? Unfortunately, it does not work for me in a completely new project with the latest version of next.js and the latest version of fluentui.
> > Do you have an example project?
> > Does the plugin work with next.config.mjs or do I have to rename the file to .js?
> 
> @amroncz Will investigate and will try to create a base stackblitz with a working setup. Currently don't have the bandwidth to look into it, but once I get a change I'll update the issue #30766. Would be nice if you could add your log to the issue I linked, thanks!

Did it :) https://github.com/microsoft/fluentui/issues/30766#issuecomment-1993685583

**Comment by amroncz** [comment]
> Hi, it isn't working for me either. I've followed the exact same steps as outlined in https://react.fluentui.dev/?path=/docs/concepts-developer-server-side-rendering-next-js-appdir-setup--page.
> 
> If any code or log is needed, please let me know.

I didn't know the guide before. I tried the same thing and it didn't work.

**Comment by aquichita** [comment]
> Hi, it isn't working for me either. I've followed the exact same steps as outlined in https://react.fluentui.dev/?path=/docs/concepts-developer-server-side-rendering-next-js-appdir-setup--page.
> 
> If any code or log is needed, please let me know.

I tried the same thing and it didn't work.

**Comment by sopranopillow** [comment]
Hello, if you are having issues with NextJS 14.1.2 > refer to this issue https://github.com/microsoft/fluentui/issues/30766#issuecomment-2017523940
